File: src/main/kotlin/com/stocks/service/RiskMetricsService.kt
package com.stocks.service

import com.stocks.model.*
import org.jetbrains.exposed.v1.core.SortOrder
import org.jetbrains.exposed.v1.core.eq
import org.jetbrains.exposed.v1.jdbc.select
import org.jetbrains.exposed.v1.jdbc.transactions.transaction
import org.jetbrains.exposed.v1.jdbc.upsert
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.YearMonth
import kotlin.math.pow

private const val MIN_RELIABLE_DATA_POINTS = 12
private const val LOOKBACK_YEARS = 5L

data class RegressionResult(
    val beta: Double,
    val alpha: Double,
    val rSquared: Double,
    val dataPoints: Int,
)

data class RiskMetricsResult(
    val ticker: String,
    val name: String?,
    val type: String?,
    val beta: Double?,
    val alphaAnnual: Double?,
    val rSquared: Double?,
    val dataPoints: Int,
    val reliable: Boolean,
)

data class RiskMetricsSummary(
    val metrics: List<RiskMetricsResult>,
    val portfolioBeta: Double?,
    val cdiAnnual: Double?,
    val calculatedAt: LocalDate?,
)

@Service
class RiskMetricsService(
    private val benchmarkService: BenchmarkService,
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    fun recalculate() {
        benchmarkService.fetchAndStoreIbovMonthly()

        val ibovPrices = benchmarkService.getMonthlyPricesMap()
        if (ibovPrices.size < 2) {
            logger.warn("Not enough IBOV data to compute risk metrics")
            return
        }

        val cdiAnnual = benchmarkService.fetchCdiAnnualRate()
        val rfMonthly = cdiAnnual?.let { (1 + it).pow(1.0 / 12.0) - 1 } ?: 0.0

        val cutoffDate = LocalDate.now().minusYears(LOOKBACK_YEARS).withDayOfMonth(1)

        val ibovReturns = computeMonthlyReturns(ibovPrices, cutoffDate)
        if (ibovReturns.size < 2) {
            logger.warn("Not enough IBOV monthly returns for regression")
            return
        }

        val assets =
            transaction {
                AssetEntity
                    .all()
                    .toList()
                    .filter { it.type in listOf("STOCK", "REIT") }
                    .filter { !it.transactions.empty() }
                    .map { Triple(it.ticker.value, it.name, it.type) }
            }

        val today = LocalDate.now()

        for ((ticker, _, _) in assets) {
            val assetMonthlyPrices = getAssetMonthlyPrices(ticker, cutoffDate)
            if (assetMonthlyPrices.size < 2) {
                storeResult(ticker, today, null, 0, cdiAnnual)
                continue
            }

            val assetReturns = computeMonthlyReturns(assetMonthlyPrices, cutoffDate)

            // Align months: only use months present in both series
            val commonMonths = assetReturns.keys.intersect(ibovReturns.keys).sorted()
            if (commonMonths.size < 2) {
                storeResult(ticker, today, null, commonMonths.size, cdiAnnual)
                continue
            }

            val excessAsset = commonMonths.map { (assetReturns[it]!! - rfMonthly) }
            val excessIbov = commonMonths.map { (ibovReturns[it]!! - rfMonthly) }

            val regression = linearRegression(excessIbov, excessAsset)
            if (regression != null) {
                // Annualize monthly alpha: (1 + alpha_monthly)^12 - 1
                val alphaAnnual = (1 + regression.alpha).pow(12.0) - 1

                storeResult(
                    ticker,
                    today,
                    RegressionResult(regression.beta, alphaAnnual, regression.rSquared, regression.dataPoints),
                    regression.dataPoints,
                    cdiAnnual,
                )
            } else {
                storeResult(ticker, today, null, commonMonths.size, cdiAnnual)
            }
        }

        logger.info("Risk metrics recalculated for ${assets.size} assets")
    }

    fun getSummary(): RiskMetricsSummary {
        val latestDate =
            transaction {
                RiskMetricEntity
                    .all()
                    .orderBy(RiskMetrics.calculatedAt to SortOrder.DESC)
                    .limit(1)
                    .firstOrNull()
                    ?.calculatedAt
            } ?: return RiskMetricsSummary(emptyList(), null, null, null)

        val entities =
            transaction {
                RiskMetricEntity
                    .find { RiskMetrics.calculatedAt eq latestDate }
                    .toList()
                    .map { RiskMetricRow(it.ticker, it.beta, it.alpha, it.rSquared, it.dataPoints, it.cdiAnnual) }
            }

        val assetInfo =
            transaction {
                AssetEntity
                    .all()
                    .toList()
                    .filter { it.type in listOf("STOCK", "REIT") }
                    .associate { it.ticker.value to Pair(it.name, it.type) }
            }

        // Get current values for portfolio weighting
        val currentValues =
            transaction {
                val tickers = entities.map { it.ticker }
                tickers.associateWith { ticker ->
                    val latestPrice =
                        PriceHistories
                            .select(PriceHistories.close)
                            .where { PriceHistories.assetId eq ticker }
                            .orderBy(PriceHistories.date, SortOrder.DESC)
                            .limit(1)
                            .firstOrNull()
                            ?.get(PriceHistories.close)

                    val qty =
                        TransactionEntity
                            .find { Transactions.assetId eq ticker }
                            .toList()
                            .sumOf { tx ->
                                if (tx.type == "BUY") tx.quantity else -tx.quantity
                            }

                    if (latestPrice != null && qty > 0) qty * latestPrice else 0.0
                }
            }

        val metrics =
            entities
                .map { row ->
                    val info = assetInfo[row.ticker]
                    RiskMetricsResult(
                        ticker = row.ticker,
                        name = info?.first,
                        type = info?.second,
                        beta = row.beta,
                        alphaAnnual = row.alpha,
                        rSquared = row.rSquared,
                        dataPoints = row.dataPoints,
                        reliable = row.dataPoints >= MIN_RELIABLE_DATA_POINTS,
                    )
                }.sortedWith(compareBy({ it.type ?: "" }, { it.ticker }))

        // Portfolio weighted beta
        val totalValue = currentValues.values.sum()
        val portfolioBeta =
            if (totalValue > 0) {
                metrics
                    .filter { it.beta != null }
                    .sumOf { (currentValues[it.ticker] ?: 0.0) / totalValue * it.beta!! }
            } else {
                null
            }

        val cdiAnnual = entities.firstOrNull()?.cdiAnnual

        return RiskMetricsSummary(metrics, portfolioBeta, cdiAnnual, latestDate)
    }

    private fun getAssetMonthlyPrices(
        ticker: String,
        cutoffDate: LocalDate
    ): Map<LocalDate, Double> =
        transaction {
            PriceHistories
                .select(PriceHistories.date, PriceHistories.close)
                .where { PriceHistories.assetId eq ticker }
                .orderBy(PriceHistories.date, SortOrder.ASC)
                .toList()
                .map { it[PriceHistories.date] to it[PriceHistories.close] }
        }.filter { it.first >= cutoffDate }
            .groupBy { YearMonth.from(it.first) }
            .mapValues { (_, entries) -> entries.maxByOrNull { it.first }!!.second }
            .mapKeys { (ym, _) -> ym.atDay(1) }

    private fun storeResult(
        ticker: String,
        date: LocalDate,
        regression: RegressionResult?,
        dataPoints: Int,
        cdiAnnual: Double?,
    ) {
        transaction {
            RiskMetrics.upsert(RiskMetrics.ticker, RiskMetrics.calculatedAt) {
                it[RiskMetrics.ticker] = ticker
                it[calculatedAt] = date
                it[beta] = regression?.beta
                it[alpha] = regression?.alpha
                it[rSquared] = regression?.rSquared
                it[RiskMetrics.dataPoints] = dataPoints
                it[RiskMetrics.cdiAnnual] = cdiAnnual
            }
        }
    }

    companion object {
        fun computeMonthlyReturns(
            prices: Map<LocalDate, Double>,
            cutoffDate: LocalDate,
        ): Map<LocalDate, Double> {
            val sorted =
                prices.entries
                    .filter { it.key >= cutoffDate }
                    .sortedBy { it.key }

            if (sorted.size < 2) return emptyMap()

            val returns = mutableMapOf<LocalDate, Double>()
            for (i in 1 until sorted.size) {
                val prev = sorted[i - 1].value
                val curr = sorted[i].value
                if (prev > 0) {
                    returns[sorted[i].key] = (curr / prev) - 1.0
                }
            }
            return returns
        }

        fun linearRegression(
            x: List<Double>,
            y: List<Double>,
        ): RegressionResult? {
            val n = x.size
            if (n < 2 || n != y.size) return null

            val xMean = x.average()
            val yMean = y.average()

            var covXY = 0.0
            var varX = 0.0
            var varY = 0.0

            for (i in 0 until n) {
                val dx = x[i] - xMean
                val dy = y[i] - yMean
                covXY += dx * dy
                varX += dx * dx
                varY += dy * dy
            }

            if (varX < 1e-12) return null

            val beta = covXY / varX
            val alpha = yMean - beta * xMean
            val rSquared = if (varY > 1e-12) (covXY * covXY) / (varX * varY) else 0.0

            return RegressionResult(
                beta = beta,
                alpha = alpha,
                rSquared = rSquared,
                dataPoints = n,
            )
        }
    }
}

private data class RiskMetricRow(
    val ticker: String,
    val beta: Double?,
    val alpha: Double?,
    val rSquared: Double?,
    val dataPoints: Int,
    val cdiAnnual: Double?,
)

File: src/modules/risk/risk-metrics.service.ts
import type { Db } from '../../config/db.js'
import {
  annualToMonthlyRate,
  computeMonthlyReturns,
  linearRegression,
  monthlyToAnnualRate,
} from '../../domain/regression.js'
import type { Logger } from '../../integrations/http.js'
import { silentLogger } from '../../integrations/http.js'
import {
  addMonths,
  firstDayOf,
  type IsoDate,
  today as todayIso,
  yearMonth,
} from '../../shared/iso-date.js'
import type { BenchmarkService } from './benchmark.service.js'

/** Porte de src/main/kotlin/com/stocks/service/RiskMetricsService.kt. */

/** Abaixo disso a regressão existe, mas não é confiável o bastante para decidir nada. */
export const MIN_RELIABLE_DATA_POINTS = 12
const LOOKBACK_YEARS = 5
const RISK_ASSET_TYPES = ['STOCK', 'REIT']

/** `with` esconde quem não tem posição aberta; `all` lista todo ativo já calculado. */
export type PositionFilter = 'with' | 'all'

export type RiskMetricsResult = {
  ticker: string
  name: string | null
  type: string | null
  hasPosition: boolean
  currentValue: number
  /** Fatia que a posição ocupa dentro do próprio tipo; null quando o tipo não vale nada. */
  typeShare: number | null
  beta: number | null
  alphaAnnual: number | null
  rSquared: number | null
  dataPoints: number
  reliable: boolean
}

export type RiskMetricsSummary = {
  metrics: RiskMetricsResult[]
  portfolioBeta: number | null
  cdiAnnual: number | null
  calculatedAt: IsoDate | null
}

export class RiskMetricsService {
  constructor(
    private readonly db: Db,
    private readonly benchmarks: BenchmarkService,
    private readonly logger: Logger = silentLogger,
  ) {}

  /** Recalcula beta, alfa e R² de cada ação e FII contra o Ibovespa. */
  async recalculate(today: IsoDate = todayIso()): Promise<void> {
    await this.benchmarks.fetchAndStoreIbovMonthly(today)

    const ibovPrices = await this.benchmarks.getMonthlyPricesMap()
    if (ibovPrices.size < 2) {
      this.logger.warn('Dados do IBOV insuficientes para calcular métricas de risco')
      return
    }

    const cdiAnnual = await this.benchmarks.fetchCdiAnnualRate()
    const rfMonthly = cdiAnnual === null ? 0 : annualToMonthlyRate(cdiAnnual)
    const cutoffDate = firstDayOf(yearMonth(addMonths(today, -12 * LOOKBACK_YEARS)))

    const ibovReturns = computeMonthlyReturns(ibovPrices, cutoffDate)
    if (ibovReturns.size < 2) {
      this.logger.warn('Retornos mensais do IBOV insuficientes para a regressão')
      return
    }

    const assets = await this.eligibleAssets()

    for (const ticker of assets) {
      const assetPrices = await this.monthlyPrices(ticker, cutoffDate)
      if (assetPrices.size < 2) {
        await this.storeResult(ticker, today, null, 0, cdiAnnual)
        continue
      }

      const assetReturns = computeMonthlyReturns(assetPrices, cutoffDate)
      // Só meses presentes nas duas séries — regressão exige pares.
      const commonMonths = [...assetReturns.keys()].filter((m) => ibovReturns.has(m)).sort()

      if (commonMonths.length < 2) {
        await this.storeResult(ticker, today, null, commonMonths.length, cdiAnnual)
        continue
      }

      const excessAsset = commonMonths.map((m) => (assetReturns.get(m) as number) - rfMonthly)
      const excessIbov = commonMonths.map((m) => (ibovReturns.get(m) as number) - rfMonthly)

      const regression = linearRegression(excessIbov, excessAsset)
      if (regression === null) {
        await this.storeResult(ticker, today, null, commonMonths.length, cdiAnnual)
        continue
      }

      await this.storeResult(
        ticker,
        today,
        {
          beta: regression.beta,
          alpha: monthlyToAnnualRate(regression.alpha),
          rSquared: regression.rSquared,
        },
        regression.dataPoints,
        cdiAnnual,
      )
    }

    this.logger.info(`Métricas de risco recalculadas para ${assets.length} ativos`)
  }

  /** Última rodada de métricas, com o beta da carteira ponderado pelo valor de cada posição. */
  async getSummary(position: PositionFilter = 'with'): Promise<RiskMetricsSummary> {
    const latest = await this.db.riskMetric.findFirst({
      orderBy: { calculatedAt: 'desc' },
      select: { calculatedAt: true },
    })
    if (latest === null) {
      return { metrics: [], portfolioBeta: null, cdiAnnual: null, calculatedAt: null }
    }

    const rows = await this.db.riskMetric.findMany({ where: { calculatedAt: latest.calculatedAt } })
    const assets = await this.db.asset.findMany({
      where: { type: { in: RISK_ASSET_TYPES } },
      select: { ticker: true, name: true, type: true, quantity: true, hasPosition: true },
    })
    const assetByTicker = new Map(assets.map((a) => [a.ticker, a]))

    const currentValues = await this.currentValues(rows.map((r) => r.ticker))

    // Somado sobre todos os ativos, mesmo os que o filtro vai esconder: quem não tem
    // posição vale zero e não muda o total do tipo.
    const totalByType = new Map<string, number>()
    for (const row of rows) {
      const type = assetByTicker.get(row.ticker)?.type
      if (type === undefined || type === null) continue
      totalByType.set(type, (totalByType.get(type) ?? 0) + (currentValues.get(row.ticker) ?? 0))
    }

    const all: RiskMetricsResult[] = rows
      // Sem regressão a linha inteira seria um traço em cada coluna — não informa nada.
      // O ativo continua contando no total do tipo e no beta da carteira, só não é listado.
      .filter((row) => row.beta !== null)
      .map((row) => {
        const asset = assetByTicker.get(row.ticker)
        const currentValue = currentValues.get(row.ticker) ?? 0
        const typeTotal = asset?.type == null ? 0 : (totalByType.get(asset.type) ?? 0)
        return {
          ticker: row.ticker,
          name: asset?.name ?? null,
          type: asset?.type ?? null,
          hasPosition: (asset?.hasPosition ?? false) && (asset?.quantity ?? 0) > 0,
          currentValue,
          typeShare: typeTotal > 0 ? currentValue / typeTotal : null,
          beta: row.beta,
          alphaAnnual: row.alpha,
          rSquared: row.rSquared,
          dataPoints: row.dataPoints,
          reliable: row.dataPoints >= MIN_RELIABLE_DATA_POINTS,
        }
      })
      .sort(
        (a, b) => (a.type ?? '').localeCompare(b.type ?? '') || a.ticker.localeCompare(b.ticker),
      )

    // O beta da carteira ignora o filtro: ele descreve a carteira inteira, não a lista à vista.
    const totalValue = [...currentValues.values()].reduce((total, v) => total + v, 0)
    const portfolioBeta =
      totalValue > 0
        ? all
            .filter((m) => m.beta !== null)
            .reduce((total, m) => total + (m.currentValue / totalValue) * (m.beta as number), 0)
        : null

    return {
      metrics: position === 'all' ? all : all.filter((m) => m.hasPosition),
      portfolioBeta,
      cdiAnnual: rows[0]?.cdiAnnual ?? null,
      calculatedAt: latest.calculatedAt as IsoDate,
    }
  }

  private async eligibleAssets(): Promise<string[]> {
    const withTransactions = await this.db.transaction.groupBy({ by: ['assetId'] })
    const tickers = new Set(withTransactions.map((t) => t.assetId))

    const assets = await this.db.asset.findMany({
      where: { type: { in: RISK_ASSET_TYPES } },
      select: { ticker: true },
      orderBy: { ticker: 'asc' },
    })
    return assets.map((a) => a.ticker).filter((ticker) => tickers.has(ticker))
  }

  /** Último preço de cada mês do ativo, a partir do histórico diário. */
  private async monthlyPrices(ticker: string, cutoffDate: IsoDate): Promise<Map<IsoDate, number>> {
    const rows = await this.db.priceHistory.findMany({
      where: { assetId: ticker, date: { gte: cutoffDate } },
      orderBy: { date: 'asc' },
      select: { date: true, close: true },
    })

    const byMonth = new Map<IsoDate, number>()
    for (const row of rows) {
      byMonth.set(firstDayOf(yearMonth(row.date as IsoDate)), row.close)
    }
    return byMonth
  }

  private async currentValues(tickers: readonly string[]): Promise<Map<string, number>> {
    const values = new Map<string, number>()
    for (const ticker of tickers) {
      const asset = await this.db.asset.findUnique({
        where: { ticker },
        select: { quantity: true, hasPosition: true },
      })
      if (asset === null || !asset.hasPosition || asset.quantity <= 0) {
        values.set(ticker, 0)
        continue
      }
      const price = await this.db.priceHistory.findFirst({
        where: { assetId: ticker },
        orderBy: { date: 'desc' },
        select: { close: true },
      })
      values.set(ticker, price === null ? 0 : asset.quantity * price.close)
    }
    return values
  }

  private async storeResult(
    ticker: string,
    date: IsoDate,
    regression: { beta: number; alpha: number; rSquared: number } | null,
    dataPoints: number,
    cdiAnnual: number | null,
  ): Promise<void> {
    const data = {
      beta: regression?.beta ?? null,
      alpha: regression?.alpha ?? null,
      rSquared: regression?.rSquared ?? null,
      dataPoints,
      cdiAnnual,
    }
    await this.db.riskMetric.upsert({
      where: { ticker_calculatedAt: { ticker, calculatedAt: date } },
      create: { ticker, calculatedAt: date, ...data },
      update: data,
    })
  }
}
